networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/core/PeerRequiredLocalMessageHandler.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.networking.eth2.rpc.core;
+
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.networking.eth2.peers.Eth2Peer;
+
+public abstract class PeerRequiredLocalMessageHandler<I, O> implements LocalMessageHandler<I, O> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  @Override
+  public void onIncomingMessage(
+      final Optional<Eth2Peer> maybePeer, final I message, final ResponseCallback<O> callback) {
+    maybePeer.ifPresentOrElse(
+        peer -> onIncomingMessage(peer, message, callback),
+        () -> {
+          LOG.trace(
+              "Ignoring message of type {} because peer has disconnected", message.getClass());
+          callback.completeWithError(
+              new RpcException(RpcResponseStatus.SERVER_ERROR_CODE, "Peer disconnected"));
```

Comment:
This isn't really a server error - what about completing with a `PeerDisconnectedException`?  And then we could update `completeWithError` to only attempt writing the message to our peer if it gets an `RcpException`.

Reply:
I like the separation of RPC error responses from unexpected errors.  We do need to send unexpected responses back as a SERVER_ERROR response but can skip doing so for PeerDisconnectedException and it would generally be nicer to push that handling into `RpcResponseCallback` instead of randomly reporting server errors from the handling code.

So I've now got two methods in `ResponseCallback` `completeWithErrorResponse(RpcException)` and `completeWithUnexpectedError(Throwable)`.